File: client/src/modules/assetmanager/schemas/financial/kpi-value.schemas.ts
/**
 * KPI Value Schemas
 *
 * Zod validation schemas for KPIValue model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/financial_models.py
 * - Schema: /server/apps/assetmanager/schemas/financial_schemas/kpi_value_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/financial_subrouters/kpi_value_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// KPIValue Schema (Full Representation)
// ==========================================

/**
 * KPIValue schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class KPIValue(BaseModel)
 */
export const KPIValueSchema = z.object({
  id: z.number(),
  kpi_id: z.number(),

  // Time Dimensions
  year: z.number(),
  quarter: z.string().nullable(),
  semester: z.string().nullable(),
  month: z.string().nullable(),
  full_year: z.boolean(),
  scenario: z.string(),
  date: z.string().nullable(),

  // Value
  value: z.number().nullable(),
  notes: z.string().nullable(),

  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(), // ISO datetime string from backend
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new KPI value (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class KPIValueCreate(BaseModel)
 */
export const CreateKPIValueSchema = z.object({
  kpi_id: z.number(),
  year: z.number(),
  full_year: z.boolean().optional(),
  scenario: z.string().optional(),
  quarter: z.string().nullable().optional(),
  semester: z.string().nullable().optional(),
  month: z.string().nullable().optional(),
  date: z.string().nullable().optional(),
  value: z.number().nullable().optional(),
  notes: z.string().nullable().optional(),
});

/**
 * Schema for updating a KPI value (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class KPIValueUpdate(BaseModel)
 */
export const UpdateKPIValueSchema = z.object({
  kpi_id: z.number().nullable().optional(),

  // Time Dimensions
  year: z.number().nullable().optional(),
  quarter: z.string().nullable().optional(),
  semester: z.string().nullable().optional(),
  month: z.string().nullable().optional(),
  full_year: z.boolean().nullable().optional(),
  scenario: z.string().nullable().optional(),
  date: z.string().nullable().optional(),

  // Value
  value: z.number().nullable().optional(),
  notes: z.string().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type KPIValue = z.infer<typeof KPIValueSchema>;
export type CreateKPIValue = z.infer<typeof CreateKPIValueSchema>;
export type UpdateKPIValue = z.infer<typeof UpdateKPIValueSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single KPI value
 * Backend equivalent: class KPIValueResponse(BaseModel)
 */
export type KPIValueResponse = {
  success: boolean;
  data?: KPIValue;
  error?: string;
};

/**
 * Response containing multiple KPI values
 * Backend equivalent: class KPIValuesResponse(BaseModel)
 */
export type KPIValuesResponse = {
  success: boolean;
  data?: KPIValue[];
  error?: string;
};

File: server/apps/assetmanager/schemas/financial_schemas/kpi_value_schemas.py
"""
KPIValue Schemas

Pydantic schemas for the KPIValue model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from datetime import date as date_type

# ==========================================
# KPIValue Schema (Full Representation)
# ==========================================

class KPIValue(BaseModel):
    """KPIValue schema - full representation"""
    id: int
    kpi_id: int = Field(description="Associated KPI ID")

    # Time Dimensions
    year: int = Field(description="Fiscal year")
    quarter: str | None = Field(None, description="Fiscal quarter")
    semester: str | None = Field(None, description="Fiscal semester")
    month: str | None = Field(None, description="Fiscal month")
    full_year: bool = Field(description="Whether this is a full year value")
    scenario: str = Field(description="Scenario type (actual, budget, forecast)")
    date: date_type | None = Field(None, description="KPI value date")

    # Value
    value: float | None = Field(None, description="KPI value")
    notes: str | None = Field(None, description="Additional notes")

    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class KPIValueCreate(BaseModel):
    """Schema for creating a new KPI value"""
    # Required fields
    kpi_id: int = Field(description="Associated KPI ID")
    year: int = Field(description="Fiscal year")

    # Fields with defaults
    full_year: bool = Field(default=False, description="Whether this is a full year value")
    scenario: str = Field(default="actual", description="Scenario type (actual, budget, forecast)")

    # Optional fields
    quarter: str | None = Field(None, description="Fiscal quarter")
    semester: str | None = Field(None, description="Fiscal semester")
    month: str | None = Field(None, description="Fiscal month")
    date: date_type | None = Field(None, description="KPI value date")
    value: float | None = Field(None, description="KPI value")
    notes: str | None = Field(None, description="Additional notes")

class KPIValueUpdate(BaseModel):
    """Schema for updating a KPI value"""
    kpi_id: int | None = None

    # Time Dimensions
    year: int | None = None
    quarter: str | None = None
    semester: str | None = None
    month: str | None = None
    full_year: bool | None = None
    scenario: str | None = None
    date: date_type | None = None

    # Value
    value: float | None = None
    notes: str | None = None

# ==========================================
# Response Types
# ==========================================

class KPIValueResponse(BaseModel):
    """Response containing a single KPI value"""
    success: bool
    data: KPIValue | None = None
    error: str | None = None

class KPIValuesResponse(BaseModel):
    """Response containing multiple KPI values"""
    success: bool
    data: list[KPIValue] | None = None
    error: str | None = None
